Includes FFTW if USE_INTEL_MKL is enabled

bmharsha opened the pull request:
Setting USE_INTEL_MKL to 1 will make sure FFTW is included in the final build (Even if Base.USE_GPL_LIBS is disabled)

aviks commented:
To clarify, this includes the the julia interface, since MKL supports the same interface but the FFTW native library is still excluded. Right?

tkelman commented:
This should just use the existing `fftw_vendor()` instead of adding a bunch of new constants

bmharsha commented:
> To clarify, this includes the the julia interface, since MKL supports the same interface but the FFTW native library is still excluded. Right?

Correct.

ranjanan commented on base/dft.jl:
```diff
@@ -580,7 +580,7 @@ module FFTW
     function plan_r2r end
 
-    Base.USE_GPL_LIBS && include(joinpath("fft", "FFTW.jl"))
+    (Base.USE_GPL_LIBS || Base.USE_INTEL_MKL_FFT) && include(joinpath("fft", "FFTW.jl"))
```
Instead of using the constants from the Makefile, you can change this check to: 

`(Base.USE_GPL_LIBS || Base.fftw_vendor() == :mkl)`

You don't need the constants in the Makefile thereafter.

ViralBShah replied:
Shouldn't we just make `fftw_vendor()` return false if there isn't one and then include the FFTW interface file as long as this function does not return false?

ViralBShah replied:
Of course there aren't that many fftw vendors out there, so perhaps no need for overkill here either.

bmharsha committed:
Includes FFTW if USE_INTEL_MKL is enabled

Setting USE_INTEL_MKL to 1 will make sure FFTW is included in the final
build (Even if Base.USE_GPL_LIBS is disabled)

tkelman commented:
Should be squashed, but this looks fine. Separately from this, if you're testing on Windows it looks like the implementation of `fftw_vendor` could probably use an extra case to account for the MKL dll being called `mkl_rt` there rather than `libmkl_rt`.

ViralBShah commented:
We should handle the windows thing.

ranjanan commented:
Should this have a back port pending 0.5 label on it?

bmharsha committed:
Adding an extra use-case to fftw_vendor function to provide support on Windows

On Windows, MKL dll is called mkl_rt, fftw_vendor() was not handling
this use-case, this commit makes sure fftw_vendor() handles this
situation and returns appropriate value.

tkelman commented on base/util.jl:
```diff
@@ -383,7 +383,7 @@ macro timed(ex)
 
 function fftw_vendor()
-    if Base.libfftw_name == "libmkl_rt"
+    if Base.libfftw_name == "libmkl_rt" || (is_windows() && Base.libfftw_name == "mkl_rt")
```
I don't think the `is_windows()` check is strictly necessary, I doubt Intel would distribute a `mkl_rt.so` or `mkl_rt.dylib` but if they did we'd likely want to count it here.

tkelman committed:
allow mkl_rt name in fftw_vendor on non windows

aviks commented:
Thanks Tony!
